File: backend/src/main/java/com/music/appmusic/entity/Song.java
package com.music.appmusic.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonManagedReference;
import jakarta.persistence.*;
import lombok.Data;

import java.util.HashSet;
import java.util.Set;

@Data
@Entity
@Table(name = "songs")
public class Song {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(nullable = false)
    private String title;

    private String coverImage;
    private String filePath;

    @Column(columnDefinition = "TEXT")
    private String lyrics;

    @ManyToOne
    @JoinColumn(name = "artist_id", nullable = false)
    @JsonManagedReference
    private Artist artist;

    @ManyToOne
    @JoinColumn(name = "album_id")
    @JsonBackReference
    private Album album;

    @ManyToOne
    @JoinColumn(name = "media_type_id", nullable = false)
    @JsonBackReference
    private MediaType mediaType;

    @ManyToMany
    @JoinTable(
            name = "user_liked_songs",
            joinColumns = @JoinColumn(name = "song_id"),
            inverseJoinColumns = @JoinColumn(name = "user_id")
    )
    private Set<User> likedByUsers = new HashSet<>();

    @Column(name = "play_count", nullable = false)
    private Long playCount = 0L;
}

File: lib/UI/models/song.dart
import 'artist.dart';

class Song {
  final int id;
  final String title;
  final String? coverImage;
  final String? filePath;
  final int? duration;
  final Artist? artist;
  final int? albumId;
  final int? mediaTypeId;
  final String? lyrics;
  final int playCount;

  Song({
    required this.id,
    required this.title,
    this.coverImage,
    this.filePath,
    this.duration,
    this.artist,
    this.albumId,
    this.mediaTypeId,
    this.lyrics,
    required this.playCount,
  });

  factory Song.fromJsonBrief(Map<String, dynamic> json) {
    Map<String, dynamic>? artistJson = json['artist'];
    Artist? artist;
    if (artistJson != null) {
      artist = Artist(
        id: artistJson['id'] as int,
        name: artistJson['name'] ?? 'Nghệ sĩ ẩn danh',
        profileImage: artistJson['profileImage'] as String?,
      );
    }

    return Song(
      id: json['id'] as int,
      title: json['title'] ?? 'Unknown',
      coverImage: json['coverImage'] as String?,
      filePath: json['filePath'] as String?,
      duration: json['duration'] as int?,
      artist: artist,
      albumId: json['albumId'] as int?,
      mediaTypeId: json['mediaTypeId'] as int?,
      lyrics: json['lyrics'] as String?,
      playCount: json['playCount'] as int? ?? 0,
    );
  }

  factory Song.fromJson(Map<String, dynamic> json) {
    return Song.fromJsonBrief(json);
  }
}